### PR TITLE
cmake: Check for C11 atomics and set config

### PR DESCRIPTION
**Pull request by cblichmann**
Context #8973

### CODE REVIEW

**Comment by cblichmann** [comment]
Since thread-safe init is a significant feature (IMO), I think it does.
I'll update the PR.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1053,6 +1053,16 @@ if(HAVE_SIZEOF_LONG_LONG)
   set(HAVE_LL 1)
 endif()
 
+# Check for atomics support
+check_c_source_compiles(
+  "
+  int main(int, char**)
+  {
+    _Atomic int i = 0;
+    return 0;
```
Comment (friendlyanon) [inline CMakeLists.txt:1062]:
```suggestion
  #include <stdatomic.h>
  
  int main(void)
  {
    atomic_int i = 0;
    return atomic_load(&i);
```

I would personally change the test to this, so there is no chance of warnings being raised here. This is enough to not make Clang emit anything with `-Weverything`.

Comment (cblichmann) [reply CMakeLists.txt:1062]: Sure. I'll change the check in the Autotools code as well then.

Comment (bagder) [reply CMakeLists.txt:1062]: is `atomic_int` really C11? I can only find references to it in C++ docs?

**Comment by cblichmann** [comment]
Any tips for how I should be dealing with the test failures? Some seem unrelated to this change and others I don't quite understand yet.

Unrelated (missing `psl` feature):
```
2022-06-08 16:31:48.687221 | ubuntu-bionic | test 1014...[Compare curl --version with curl-config --features]
2022-06-08 16:31:49.525915 | ubuntu-bionic | Mismatch in features lists:
2022-06-08 16:31:49.525971 | ubuntu-bionic | curl:        alt-svc asynchdns brotli hsts http3 https-proxy idn ipv6 largefile ntlm psl ssl threadsafe-init unixsockets zstd
2022-06-08 16:31:49.525984 | ubuntu-bionic | curl-config: alt-svc asynchdns brotli hsts http3 https-proxy idn ipv6 largefile ntlm ssl threadsafe-init unixsockets zstd
2022-06-08 16:31:49.526489 | ubuntu-bionic |  postcheck FAILED
```

Failing on Windows using msys v2:
```
2022-06-08T18:00:16.9622957Z test 3026...[curl_global_init thread-safety]
2022-06-08T18:00:17.6232259Z 
2022-06-08T18:00:17.6233251Z lib3026.exe returned 127, when expecting 0
2022-06-08T18:00:17.6233869Z  exit FAILED
2022-06-08T18:00:17.6238697Z == Contents of files in the log/ dir after test 3026
2022-06-08T18:00:17.6245538Z === Start of file commands.log
2022-06-08T18:00:17.6249392Z  ./libtest/lib3026.exe none >log/stdout3026 2>log/stderr3026
2022-06-08T18:00:17.6250062Z === End of file commands.log
2022-06-08T18:00:17.6255678Z === Start of file ftpserver.cmd
2022-06-08T18:00:17.6259999Z  Testnum 3026
2022-06-08T18:00:17.6260525Z === End of file ftpserver.cmd
2022-06-08T18:00:17.6266788Z === Start of file stderr3026
2022-06-08T18:00:17.6271480Z  URL: none
2022-06-08T18:00:17.6283756Z  lib3026.c:65 Couldn't create thread, errno 11
2022-06-08T18:00:17.6284691Z === End of file stderr3026
```

Shouldn't the above also fail when we're preparing the build files with autotools?

**Comment by bagder** [comment]
The issue here is that the `threadsafe-init` feature needs to be enabled for curl-config for Windows as well, and not only for `_Atomic` present. Like is done in this configure check:

https://github.com/curl/curl/blob/5121d98c22dc077d72c68e7b9436a04f55bd7249/configure.ac#L4261-L4276

**Comment by Keitagit-kun** [comment]
So, if target _WIN32_WINNT < 0x600 threadsafe-init feature should be disabled (will be not selectable)?

**Comment by bagder** [comment]
Exactly. On Windows we rely on `SWRLOCK` instead of atomic and then presumably that is the lowest Windows version that supports it. See [easy_lock.h](https://github.com/curl/curl/blob/master/lib/easy_lock.h)

**Comment by bagder** [comment]
This [fails on macos](https://github.com/curl/curl/runs/6834074919?check_suite_focus=true#step:6:9):
~~~
Run cmake -H. -Bbuild -DCURL_WERROR=ON -DPICKY_COMPILER=ON -DOPENSSL_ROOT_DIR=/usr/local/opt/libressl -DCURL_DISABLE_LDAP=ON -DCURL_DISABLE_LDAPS=ON
CMake Error at CMakeLists.txt:1080:
  Parse error.  Expected "(", got newline with text "
  ".
-- Configuring incomplete, errors occurred!

**Comment by cblichmann** [comment]
Sorry about that, I missed the parenthesis afer `endif`. Apparently some CMake version simply accept this (as did my local one).

**Comment by cblichmann** [comment]
According to #8996, we can ignore test 3026 failing.

**Comment by cblichmann** [comment]
Let's do the fix for `-H` in a different PR :)

File: CMakeLists.txt [hunk 2]
```diff
@@ -1464,6 +1489,7 @@ _add_if("HTTP3"         USE_NGTCP2 OR USE_QUICHE)
 _add_if("MultiSSL"      CURL_WITH_MULTI_SSL)
 _add_if("HTTPS-proxy"   SSL_ENABLED AND (USE_OPENSSL OR USE_GNUTLS OR USE_NSS))
 _add_if("unicode"       ENABLE_UNICODE)
+_add_if("threadsafe-init" (NOT WIN32 AND HAVE_ATOMIC) OR (HAVE_SRW_LOCK))
```
Comment (bagder) [inline CMakeLists.txt:1492]: This feature has been renamed to `threadsafe` while this PR has been in the works!

**Comment by bagder** [comment]
Can you figure out why the appveyor windows builds using cmake don't get the `threadsafe` feature listed in `curl-config` output thus making test 1014 fail?

[example build](https://ci.appveyor.com/project/curlorg/curl/builds/43856658/job/fal7pjkuelama6di#L3616)

**Comment by cblichmann** [comment]
Looking into this. I first rebased on `master` though.

**Review by vszakats — APPROVED** [review]

**Comment by bagder** [comment]
The cmake build still [fails test 1014 on appveyor](https://ci.appveyor.com/project/curlorg/curl/builds/43865783/job/hmw0924i9qu7so2l#L3609). Which seems to be because `curl-config --feature` doesn't get the `threadsafe` feature correctly.

**Comment by bagder** [comment]
This needs a rebase and force-push. I think the curl-functions edit can be removed and the cmake check should make sure to also check for the presence of the `stdatomic.h`  header file before it defines `HAVE_ATOMIC`, to match how we improved the autotools check.

**Pull request closed without merging**